crates/typed-store/src/metrics.rs

```diff
@@ -1,15 +1,261 @@
 };
+use std::cell::RefCell;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+use rocksdb::{perf::set_perf_stats, PerfContext, PerfMetric, PerfStatsLevel};
+
+thread_local! {
+    static PER_THREAD_ROCKS_PERF_CONTEXT: std::cell::RefCell<PerfContext>  = RefCell::new(PerfContext::default());
+}
+
+#[derive(Debug, Default, Clone)]
+/// A struct for keeping track of the sampling
+pub struct PerfSampler {
+    // Number of operations since last sample
+    pub num_ops: Arc<AtomicU64>,
+    // Timestamp of the last sample
+    pub latest_sample_ts: Arc<AtomicU64>,
+}
+
+#[derive(Debug, Clone)]
+// A struct for sampling based on number of operations or duration.
+// Sampling happens if the duration expires or number of operations
+// is more than the threshold
+pub struct SamplingInterval {
+    // Sample once every time duration
+    pub once_every_duration: Duration,
+    // Sample once every number of operations
+    pub after_num_ops: u64,
+    // Perf sampler for keeping track of previous sample
+    pub perf_sampler: PerfSampler,
+}
+
+impl Default for SamplingInterval {
+    fn default() -> Self {
+        SamplingInterval {
+            once_every_duration: Duration::from_secs(10),
+            after_num_ops: u64::MAX,
+            perf_sampler: PerfSampler {
+                num_ops: Arc::new(AtomicU64::new(0)),
+                latest_sample_ts: Arc::new(AtomicU64::new(0)),
+            },
+        }
+    }
+}
+
+impl SamplingInterval {
+    pub fn new(once_every_duration: Duration, after_num_ops: u64) -> Self {
+        SamplingInterval {
+            once_every_duration,
+            after_num_ops,
+            perf_sampler: PerfSampler {
+                num_ops: Arc::new(AtomicU64::new(0)),
+                latest_sample_ts: Arc::new(AtomicU64::new(0)),
+            },
+        }
+    }
+    pub fn sample(&self) -> bool {
+        if self.after_num_ops == u64::MAX && self.once_every_duration == Duration::MAX {
+            return false;
+        }
+        if self.after_num_ops != u64::MAX {
+            let num_ops = self.perf_sampler.num_ops.fetch_add(1, Ordering::Relaxed);
+            if num_ops >= self.after_num_ops {
+                if num_ops > self.after_num_ops {
+                    // Multiple threads got here at the same time, only one
+                    // should be able to get the sample
+                    return false;
+                }
+                let now = Clock::recent_since_epoch();
+                self.perf_sampler.num_ops.store(0, Ordering::SeqCst);
+                self.perf_sampler
+                    .latest_sample_ts
+                    .store(now.as_secs(), Ordering::SeqCst);
+                return true;
+            }
+        }
+        if self.once_every_duration == Duration::MAX {
+            return false;
+        }
+        let now = Clock::recent_since_epoch();
+        let prev_ts = self.perf_sampler.latest_sample_ts.load(Ordering::Relaxed);
```

Comment:
Hmm, so this PerfSampler seems to have been brought over from the other PR, but it seems lots of optimizations went into this and this seems to be an earlier rev of the design?   Cuz this still has an Atomic for the prev_ts and didn't seem to have a bunch of other changes?

Reply:
Just rebased, and it should look better now